server/test/services/program/predicate/PredicateGeneratorTest.java

```diff
@@ -0,0 +1,316 @@
+    assertThat(predicateDefinition.rootNode())
+        .isEqualTo(
+            PredicateExpressionNode.create(
+                LeafOperationExpressionNode.builder()
+                    .setQuestionId(testQuestionBank.applicantAddress().id)
+                    .setScalar(Scalar.ZIP)
+                    .setOperator(Operator.EQUAL_TO)
+                    .setComparedValue(PredicateValue.of("98144"))
+                    .build()));
+  }
+
+  @Test
+  public void generatePredicateDefinition_multiQuestion_multiValue() throws Exception {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "predicateAction",
+                "SHOW_BLOCK",
+                String.format("question-%d-scalar", testQuestionBank.applicantAddress().id),
+                "ZIP",
+                String.format("question-%d-operator", testQuestionBank.applicantAddress().id),
+                "EQUAL_TO",
+                String.format(
+                    "group-1-question-%d-predicateValue", testQuestionBank.applicantAddress().id),
+                "98144",
+                String.format(
+                    "group-2-question-%d-predicateValue", testQuestionBank.applicantAddress().id),
+                "10001",
+                String.format("question-%d-scalar", testQuestionBank.applicantDate().id),
+                "DATE",
+                String.format("question-%d-operator", testQuestionBank.applicantDate().id),
+                "EQUAL_TO",
+                String.format(
+                    "group-1-question-%d-predicateValue", testQuestionBank.applicantDate().id),
+                "2023-01-01",
+                String.format(
+                    "group-2-question-%d-predicateValue", testQuestionBank.applicantDate().id),
+                "2023-02-02"));
+
+    PredicateDefinition predicateDefinition =
+        predicateGenerator.generatePredicateDefinition(form, readOnlyQuestionService);
+
+    assertThat(predicateDefinition.predicateFormat().get())
+        .isEqualTo(PredicateDefinition.PredicateFormat.OR_OF_SINGLE_LAYER_ANDS);
+    assertThat(predicateDefinition.action()).isEqualTo(PredicateAction.SHOW_BLOCK);
+    assertThat(predicateDefinition.getQuestions())
+        .containsExactlyInAnyOrder(
+            testQuestionBank.applicantAddress().id, testQuestionBank.applicantDate().id);
+
+    assertThat(predicateDefinition.rootNode().getType()).isEqualTo(PredicateExpressionNodeType.OR);
+    assertThat(predicateDefinition.rootNode().getOrNode().children())
+        .containsExactlyInAnyOrder(
+            PredicateExpressionNode.create(
+                AndNode.create(
+                    ImmutableList.of(
+                        PredicateExpressionNode.create(
+                            LeafOperationExpressionNode.builder()
+                                .setQuestionId(testQuestionBank.applicantDate().id)
+                                .setScalar(Scalar.DATE)
```

Comment:
It seems incorrect that the input order was ZIP then DATE but the verification order is DATE then ZIP.  This looks like alpha order and not the user order.  Though It's unclear the the data ordering is the correct approach either.

Reply:
This one you can ignore from our convo I think.